packages/notifier/src/publish-kit.js

```diff
@@ -180,12 +200,221 @@ export const makePublishKit = () => {
+ * associated with a given state object backing a durable publish kit.
+ * They are lost on upgrade, but recreated on-demand.
+ * Such recreation preserves the value in (but not the identity of) the
+ * current { value, done } result when possible, which is always the
+ * case when that value is terminal (i.e., from `finish` or `fail`) or
+ * when the durable publish kit is configured with
+ * `valueDurability: 'mandatory'`.
+ *
+ * @param {DurablePublishKitState} state
+ */
+const provideDurablePublishKitEphemeralData = state => {
+  const foundData = durablePublishKitEphemeralData.get(state);
+  if (foundData) {
+    return foundData;
+  }
+  const { status, publishCount } = state;
+  let newData;
+  if (status === 'failed') {
+    newData = {
+      currentP: makeQuietRejection(state.value),
+      tailP: makeQuietRejection(
+        new Error('Cannot read past end of iteration.'),
+      ),
+      tailR: undefined,
+    };
+  } else if (status === 'finished') {
+    // XXX Should we be harden()ing rejection-reason errors?
+    const tailP = makeQuietRejection(
+      new Error('Cannot read past end of iteration.'),
+    );
+    newData = {
+      currentP: HandledPromise.resolve(
+        harden({
+          head: { value: state.value, done: true },
+          publishCount,
+          tail: tailP,
+        }),
+      ),
+      tailP,
+      tailR: undefined,
+    };
+  } else if (status === 'live') {
+    const { promise: tailP, resolve: tailR } = makePromiseKit();
+    newData = {
+      currentP: state.hasValue
+        ? HandledPromise.resolve(
+            harden({
+              head: { value: state.value, done: false },
+              publishCount,
+              tail: tailP,
+            }),
+          )
+        : tailP,
+      tailP,
+      tailR,
+    };
+  } else {
+    assert.fail(X`Invalid durable promise kit status: ${q(status)}`);
+  }
+  durablePublishKitEphemeralData.set(state, harden(newData));
+  return newData;
+};
+
+/**
+ * Extends the sequence of results.
+ *
+ * @param {DurablePublishKitState} state
+ * @param {boolean} done
+ * @param {any} value
+ * @param {any} [rejection]
+ */
+const advanceDurablePublishKit = (state, done, value, rejection) => {
+  const { valueDurability, status } = state;
+  if (status !== 'live') {
+    throw new Error('Cannot update state after termination.');
+  }
+  if (done || valueDurability === 'mandatory') {
+    canBeDurable(value) ||
+      assert.fail(X`Cannot accept non-durable value: ${value}`);
+    canBeDurable(rejection) ||
+      assert.fail(X`Cannot accept non-durable rejection: ${rejection}`);
+  }
+  const { tailP: currentP, tailR: resolveCurrent } =
+    provideDurablePublishKitEphemeralData(state);
+
+  state.publishCount += 1n;
+  const publishCount = state.publishCount;
+  let tailP;
+  let tailR;
+
+  if (done) {
+    state.status = rejection ? 'failed' : 'finished';
+    tailP = makeQuietRejection(new Error('Cannot read past end of iteration.'));
```

Comment:
Isn't `advanceDurablePublishKit` trying to add new state? Shouldn't this be
```suggestion
    tailP = makeQuietRejection(new Error('Cannot append past end of iteration.'));
```

Reply:
No, this is the rejection that will be seen by consumers attempting to read a new value after receiving a terminal one.